calico.go

```diff
@@ -0,0 +1,366 @@
+// Copyright 2015 Tigera Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"runtime"
+
+	"github.com/vishvananda/netlink"
+
+	"flag"
+
+	"github.com/containernetworking/cni/pkg/ip"
+	"github.com/containernetworking/cni/pkg/ipam"
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	. "github.com/projectcalico/calico-cni/utils"
+	"github.com/projectcalico/libcalico/lib"
+)
+
+var hostname string
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+
+	hostname, _ = os.Hostname()
+}
+
+func cmdAddK8s(args *skel.CmdArgs, k8sArgs K8sArgs, conf NetConf, theEndpoint *libcalico.Endpoint) (*types.Result, *libcalico.Endpoint, error) {
+	var err error
+	var result *types.Result
+
+	profileID := fmt.Sprintf("k8s_ns.%s", k8sArgs.K8S_POD_NAMESPACE)
+
+	if theEndpoint != nil {
+		// This happens when Docker or the node restarts. K8s calls CNI with the same details as before.
+		// Do the networking but no etcd changes should be required.
+		// There's an existing endpoint - no need to create another. Find the IP address from the endpoint
+		// and use that in the response.
+		result, err = createResultFromIP(theEndpoint.IPv4Nets[0])
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// We're assuming that the endpoint is fine and doesn't need to be changed.
+		// However, the veth does need to be recreated since the namespace has been lost.
+		_, err := DoNetworking(args, conf, result)
+		if err != nil {
+			return nil, nil, err
+		}
+		// TODO - what if labels changed during the restart?
+
+	} else {
+		// There's no existing endpoint, so we need to do the following:
+		// 1) Call the configured IPAM plugin to get IP address(es)
+		// 2) Create the veth, configuring it on both the host and container namespace.
+		// 3) Configure the calico endpoint
+
+		// 1) run the IPAM plugin and make sure there's an IPv4 address
+		result, err = ipam.ExecAdd(conf.IPAM.Type, args.StdinData)
```

Comment:
We should talk about whether we're going to continue supporting the "all-in-one" plugin/ipam binary or if we're going to start making people download two binaries.